File: companion-daemon/internal/devicetrust/auth_transcript.go
package devicetrust

import (
	"encoding/binary"
	"fmt"
)

// AuthTranscript builds the canonical domain-separated transcript that both the
// host and the device sign during M2.5-3 challenge authentication.  Every field
// is length-prefixed so the transcript has unambiguous boundaries — the signer
// never signs a JSON encoding or an unterminated concatenation.
//
// Layout:
//
//	"pokit-device-auth-v1"
//	length(role)      || role        ("host" or "device")  ← first field after domain
//	length(hostId)    || hostId
//	length(deviceId)  || deviceId
//	length(bootId)    || daemonBootId
//	length(chalId)    || challengeId bytes
//	length(clientNC)  || clientNonce bytes
//	length(serverNC)  || serverNonce bytes
//	length(createdAt) || 8-byte big-endian UTC unix milliseconds
//	length(expiresAt) || 8-byte big-endian UTC unix milliseconds
//
// The distinct role field prevents a reflected host signature from being
// replayed as a device signature, or vice versa.
type AuthTranscript struct {
	Role         string // "host" or "device"
	HostID       string
	DeviceID     string
	DaemonBootID string
	ChallengeID  []byte
	ClientNonce  []byte
	ServerNonce  []byte
	CreatedAtMS  int64 // UTC unix milliseconds (issuedAt)
	ExpiresAtMS  int64 // UTC unix milliseconds
}

// Build returns the canonical binary transcript.
func (t AuthTranscript) Build() []byte {
	putField := func(b []byte, s string) []byte {
		b = binary.BigEndian.AppendUint32(b, uint32(len(s)))
		return append(b, s...)
	}
	putBytes := func(b []byte, d []byte) []byte {
		b = binary.BigEndian.AppendUint32(b, uint32(len(d)))
		return append(b, d...)
	}

	b := []byte("pokit-device-auth-v1")
	b = putField(b, t.Role)
	b = putField(b, t.HostID)
	b = putField(b, t.DeviceID)
	b = putField(b, t.DaemonBootID)
	b = putBytes(b, t.ChallengeID)
	b = putBytes(b, t.ClientNonce)
	b = putBytes(b, t.ServerNonce)
	b = appendUint64Field(b, t.CreatedAtMS)
	b = appendUint64Field(b, t.ExpiresAtMS)
	return b
}

func appendUint64Field(b []byte, v int64) []byte {
	buf := make([]byte, 8)
	binary.BigEndian.PutUint64(buf, uint64(v))
	b = binary.BigEndian.AppendUint32(b, 8)
	return append(b, buf...)
}

// Validate returns an error if required fields are missing or out of bounds.
func (t AuthTranscript) Validate() error {
	if t.Role != "host" && t.Role != "device" {
		return fmt.Errorf("invalid role: %q", t.Role)
	}
	if t.HostID == "" || t.DeviceID == "" || t.DaemonBootID == "" {
		return fmt.Errorf("missing identity field")
	}
	if len(t.ChallengeID) != 32 || len(t.ClientNonce) != 32 || len(t.ServerNonce) != 32 {
		return fmt.Errorf("invalid nonce/challenge length")
	}
	if t.CreatedAtMS <= 0 || t.ExpiresAtMS <= 0 {
		return fmt.Errorf("invalid timestamps")
	}
	return nil
}

File: mobile/src/lib/authTranscript.ts
// M3-auth-3A: byte-exact canonical authentication transcript.
//
// Mirror of companion-daemon/internal/devicetrust/auth_transcript.go
// AuthTranscript.Build(). Layout:
//
//   "pokit-device-auth-v1"
//   length(role)      || role        ("device" or "host")
//   length(hostId)    || hostId
//   length(deviceId)  || deviceId
//   length(bootId)    || daemonBootId
//   length(chalId)    || challengeId (32 raw bytes)
//   length(clientNC)  || clientNonce (32 raw bytes)
//   length(serverNC)  || serverNonce (32 raw bytes)
//   length(8)         || createdAtMS (8-byte big-endian uint64 ms)
//   length(8)         || expiresAtMS (8-byte big-endian uint64 ms)
//
// Every field is length-prefixed (uint32-BE). The distinct role field prevents
// a host signature from being replayed as a device signature.

export interface AuthTranscriptParams {
  role: 'device' | 'host';
  hostId: string;
  deviceId: string;
  daemonBootId: string;
  challengeId: Uint8Array;   // exactly 32 bytes
  clientNonce: Uint8Array;   // exactly 32 bytes
  serverNonce: Uint8Array;   // exactly 32 bytes
  createdAtMS: number;       // UTC unix milliseconds (integer)
  expiresAtMS: number;       // UTC unix milliseconds (integer)
}

// putField appends a length-prefixed UTF-8 string field (uint32-BE length).
function putField(b: number[], s: string): void {
  const encoded = new TextEncoder().encode(s);
  putUint32(b, encoded.length);
  for (let i = 0; i < encoded.length; i++) b.push(encoded[i]);
}

// putBytes appends a length-prefixed raw byte field (uint32-BE length).
function putBytes(b: number[], d: Uint8Array): void {
  putUint32(b, d.length);
  for (let i = 0; i < d.length; i++) b.push(d[i]);
}

// putUint64Field appends length(8) then 8-byte big-endian ms timestamp.
function putUint64Field(b: number[], v: number): void {
  b.push(0, 0, 0, 8); // length prefix uint32-BE = 8
  // 8-byte big-endian uint64
  const hi = Math.floor(v / 0x100000000);
  const lo = v >>> 0;
  b.push((hi >>> 24) & 0xff, (hi >>> 16) & 0xff, (hi >>> 8) & 0xff, hi & 0xff);
  b.push((lo >>> 24) & 0xff, (lo >>> 16) & 0xff, (lo >>> 8) & 0xff, lo & 0xff);
}

function putUint32(b: number[], v: number): void {
  b.push((v >>> 24) & 0xff, (v >>> 16) & 0xff, (v >>> 8) & 0xff, v & 0xff);
}

export function buildAuthTranscript(p: AuthTranscriptParams): Uint8Array {
  const b: number[] = [];
  // Domain separator prefix bytes.
  const domain = new TextEncoder().encode('pokit-device-auth-v1');
  for (let i = 0; i < domain.length; i++) b.push(domain[i]);

  putField(b, p.role);
  putField(b, p.hostId);
  putField(b, p.deviceId);
  putField(b, p.daemonBootId);
  putBytes(b, p.challengeId);
  putBytes(b, p.clientNonce);
  putBytes(b, p.serverNonce);
  putUint64Field(b, p.createdAtMS);
  putUint64Field(b, p.expiresAtMS);

  return new Uint8Array(b);
}
